### ISSUE TITLE
Request Server folders displayed as files

### ISSUE THREAD

**Issue by kenXengineering** [pkg/sftp#268]
I'm attempting to create a request server with an S3 backend.  So far I have it listing files, downloading, and uploading, however I'm having an issue where folders in S3 are showing up as files in SFTP.  When I create the list of `os.FileInfo` I am setting the entries for S3 folders to true.  Here is the S3FileInfo object I use to hold the S3 info:

```go
type S3FileInfo struct {
	name    string
	size    int64
	modTime time.Time
	isDir   bool
}

func (s S3FileInfo) Name() string { return filepath.Base(s.name) }
func (s S3FileInfo) Size() int64  { return s.size }
func (s S3FileInfo) Mode() os.FileMode {
	ret := os.FileMode(0644)
	if s.isDir {
		ret = os.FileMode(0755) | os.ModeDir
	}
	spew.Dump(ret)
	return ret
}
func (s S3FileInfo) ModTime() time.Time { return s.modTime }
func (s S3FileInfo) IsDir() bool        { return s.isDir }
func (s S3FileInfo) Sys() interface{}   { return nil }
```
And the section of code where I set isDir:
```go
// Populate our mod time
fileInfo := &S3FileInfo{
	modTime: *obj.LastModified,
}
// If the file name is .dir, then this is a directory.
if strings.Contains(key, dirFile) || key[len(key)-1:] == "/" {
	logger.WithField("key", key).Debug("Setting isDir true")
	fileInfo.isDir = true
	// Remove the .dir from the file name
	fileInfo.name = strings.Replace(key, dirFile, "", 1)
} else {
	// Normal file
	fileInfo.name = key
	fileInfo.size = *obj.Size
}
output = append(output, fileInfo)
```

When I connect via SFTP, I see `/` (the root folder) as a file, and a directory I created `test` as a file as well.

Other issue I am seeing is my file listing for the other files is missing the size, but has the mod time.  I feel like I'm missing something simple here......

**Comment by kenXengineering** [comment]
Good point on the test for `.dir`.  Thats mainly there since S3 doesn't support empty folders easily, so my initial thought was to just drop a `.dir` file in there to make it as a directory.  I cleaned that up by spitting it with `filepath`.

How about some more code!  Here is the `Filelist` function that handles `Stat` and `List`.

```go

func (fs *S3Handler) Filelist(r *sftp.Request) (sftp.ListerAt, error) {
	logger := fs.log.WithFields(logrus.Fields{"function": "Filelist", "requst": r})
	logger.Debug("Entered Filelist")

	path := r.Filepath

	switch r.Method {
	case "List":
		logger.WithField("filePath", path).Debug("Received List Request")
		// Get the file list form S3.  We will get the complete file list for all files in the S3 Bucket with the
		// path as the S3 Prefix.
		files, err := fs.s3Client.Listfiles(path, fs.User.Directory, false)
		if err != nil {
			return nil, errors.Wrap(err, "Error listing files")
		}
		list := listerAt(files)
		spew.Dump(list)
		return list, nil
	case "Stat":
		logger.WithField("filePath", path).Debug("Received Stat Request")
		files, err := fs.s3Client.Listfiles(path, fs.User.Directory, true)
		if err != nil {
			logger.Error(err)
			return nil, errors.Wrap(err, "Error listing files")
		}
		list := listerAt(files)
		spew.Dump(list)
		return list, nil
	case "Readlink":
		return nil, errors.New("Readlink is not supported")
	}

	return nil, nil
}
```

Here is the function that gets the list of files from S3 and returns the slice of `os.FileInfo` along with the definition of `S3FileInfo` again:

```go
type S3FileInfo struct {
	name    string
	size    int64
	modTime time.Time
	isDir   bool
}

func (s S3FileInfo) Name() string { return filepath.Base(s.name) }
func (s S3FileInfo) Size() int64  { return s.size }
func (s S3FileInfo) Mode() os.FileMode {
	ret := os.FileMode(0644)
	if s.isDir {
		ret = os.FileMode(0755) | os.ModeDir
	}
	spew.Dump(ret)
	return ret
}
func (s S3FileInfo) ModTime() time.Time { return s.modTime }
func (s S3FileInfo) IsDir() bool        { return s.isDir }
func (s S3FileInfo) Sys() interface{}   { return nil }

func (s *S3Client) Listfiles(path string, prefix string, stat bool) ([]os.FileInfo, error) {
	logger := s.log.WithFields(logrus.Fields{"function": "Listfiles", "path": path, "prefix": prefix})
	logger.Debug("Entered Listfiles")
	origPath := path
	if prefix != "" {
		path = filepath.Join(prefix, path)
	}

	if s.cfg.S3Prefix != "" {
		path = filepath.Join(s.cfg.S3Prefix, path)
	}
	logger.WithField("path", path).Debug("Listing Files")
	output := make([]os.FileInfo, 0)
	marker := ""
	for {
		listInput := &s3.ListObjectsInput{
			Bucket: &s.cfg.S3Bucket,
			Prefix: &path,
		}
		if marker != "" {
			listInput.Marker = &marker
		}
		listOutput, err := s.svc.ListObjects(listInput)
		if err != nil {
			return nil, errors.Wrap(err, "Error listing files")
		}
		// Loop through the contents.
		for _, obj := range listOutput.Contents {
			// Remove the prefix if we have one
			key := *obj.Key
			if s.cfg.S3Prefix != "" {
				key = strings.Replace(key, s.cfg.S3Prefix, "", 1)
			}
			// If we were passed a prefix, strip it
			if prefix != "" {
				key = strings.Replace(key, prefix, "", 1)
			}
			// Build our S3 File Info
			var fileInfo *S3FileInfo
			// If the file name is .dir or empty then it is a directory
			_, fileName := filepath.Split(key)
			if fileName == dirFile || fileName == "" {
				// Is a directory
				logger.WithField("key", key).Debug("Setting isDir true")
				fileInfo = &S3FileInfo{
					name:    strings.Replace(key, dirFile, "", 1),
					isDir:   true,
					modTime: *obj.LastModified,
				}
			} else {
				// Is a file
				fileInfo = &S3FileInfo{
					name:    strings.Replace(key, dirFile, "", 1),
					size:    *obj.Size,
					modTime: *obj.LastModified,
				}
			}

			// If we are stating a file, then we just want to return it
			if stat && key == origPath {
				output = append(output, fileInfo)
				break
			} else {
				output = append(output, fileInfo)
			}
		}
		if *listOutput.IsTruncated {
			marker = *listOutput.Marker
			continue
		}
		break
	}
	logger.Debug("Finished Listing")
	return output, nil
}
```

To prove that `isDir` and `Size` is getting set, here is the list of `os.FileInfo` that is returned by `Filelist`

```
DEBU[0020] Finished Listing                              function=Listfiles path=/ prefix=/kherner service=S3
(pkg.listerAt) (len=3 cap=4) {
 (*pkg.S3FileInfo)(0xc4200b8b80)({
  name: (string) (len=1) "/",
  size: (int64) 0,
  modTime: (time.Time) 2018-08-30 18:45:23 +0000 UTC,
  isDir: (bool) true
 }),
 (*pkg.S3FileInfo)(0xc4200b8bc0)({
  name: (string) (len=8) "/test.go",
  size: (int64) 196,
  modTime: (time.Time) 2018-08-31 15:47:23 +0000 UTC,
  isDir: (bool) false
 }),
 (*pkg.S3FileInfo)(0xc4200b8c00)({
  name: (string) (len=6) "/test/",
  size: (int64) 0,
  modTime: (time.Time) 2018-08-30 19:29:20 +0000 UTC,
  isDir: (bool) true
 })
}
```

**Comment by kenXengineering** [comment]
Alright, I finally figured it out.  My issue is I was using the same function for both `Stat` and `List`, however I did not handle the fact that `Stat` and `List` are different.  The main problem was how I was handling the file path.

When the client `cd`s into `/test_dir`, the `Stat` command expects back one result that is the file info for the directory.  `List` then expects the content of the directory, EXCLUDING the directory information.  Originally I was returning a full list for `Stat`, then returning the listed directory as part of the results for `List`, which confuses the clients.

The _simple_ check I added was if we are not `Stat`ing, the current S3 object is technically a directory, and the S3 directory path equals the directory path we are listing, do not return it in the list!  SIMPLE!

Thanks @puellanivis for taking a look at this.  Pointing out how I handled `.dir` pointed me in the right direction.
